### PR TITLE
[10.0][IMP] purchase_stock_picking_returm_invoicing: Improve perf for qty_returned field

### PR DESCRIPTION
**Pull request by pedrobaeza**
This field is computed very slow with previous code. I have improved the speed through several techniques like read_group and shared prefetching. It might be less readable, but incredibly faster compared with the previous code.

cc @Tecnativa

### CODE REVIEW

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (yajo) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: I think you can put here directly `('origin_returned_move_id.purchase_line_id', 'in', self.ids)` and remove the previous search.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (yajo) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: I think that if you want to prefetch those UoM, you have to browse all of them first in the same query.

i.e., add this at the top of the method:

```python
self.mapped("product_uom")
```

Then you can use here `ProductUom.browse(g['product_uom'][0]).with_prefetch(self._prefetch)`

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (yajo) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: Did you mean `line.qty_returned = ...`?

**Review by yajo — COMMENTED** [review]
Please add some comments in code, it's hard to read... :confused:

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (yajo) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: Remove the brackets, you don't need a list if you have a generator :wink::

```python
uom_ids = set(g['product_uom'][0] for g in groups)
```

**Review by yajo — APPROVED** [review]

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (carlosdauden) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: Can you check this option?

        p = self._prefetch
        group_dict = defaultdict(list)
        uom_ids = set()
        for g in groups:
            group_dict[g['origin_returned_move_id'][0]].append(
                g['product_uom'][0], g['product_uom_qty']
            )
            uom_ids.add(g['product_uom'][0])
        ProductUom.browse(list(uom_ids)).with_prefetch(p)
        for line in self:
            qty = 0.0
            for gl in group_dict[line.id]:
                if gl[0] == line.product_uom.id:
                    qty += gl[1]
                else:
                    uom = ProductUom.browse(gl[0]).with_prefetch(p)
                    qty += line.product_uom
                    x = uom._compute_quantity(gl[1], line.product_uom)
            line.qty_returned = qty

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (carlosdauden) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: You can use prefetch as paremeter:
`ProductUom.browse(list(uom_ids), prefetch=p)`
https://github.com/odoo/odoo/blob/020906659e55bd41a07ab0e4e79da903af7b50c1/odoo/models.py#L4324

In the below browses, too.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (carlosdauden) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: Add condition `uom == move.purchase_line_id.product_uom` to set g['product_uom_qty'] directly instead of call _compute_quantity function

**Review by carlosdauden — CHANGES_REQUESTED** [review]

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (pedrobaeza) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py]: The same is evaluated on the `_compute_quantity` method, so no gain adding that lines

Comment (carlosdauden) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py]: I don't see it:
https://github.com/odoo/odoo/blob/3112cd2dbe7459739f00eb775d5510f0aefcb3c2/addons/product/models/product_uom.py#L91-L106

**Review by carlosdauden — APPROVED** [review]

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py:133]: Use [snake_case for variables names](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#variable-names) and [CamelCase for class names](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#odoo-python-classes).

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py:141]: `p = self._prefetch` is using the same pointer variable.
if you use `id(p)` and `id(self._prefetch)` you will the same id result.

What about uses the same `self` without extra `p` variable and without extra `with_prefetch(p)`?

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: Avoid iteration 2 times for the same `groups` variable in order to get the `origin_move_ids` and `uom_ids` at same time.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py]: Don't compute_quantity if it is the same uom

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py]: Use a `collections.defaultdict` instead.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [inline purchase_stock_picking_return_invoicing/models/purchase_order.py:164]: You can avoid use this extra loop if you use in the main loop:
`move.purchase_line_id['qty_returned']  += value`

since that is sharing the same object

Review the following example:
 - https://github.com/odoo/odoo/pull/25280/commits/0f237f129e369b0edbab83209ce6bfd58436060d

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (yajo) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:141]: He can use `self._prefetch` directly without `p`, but he needs to add `.with_prefetch()` after a browse call if he wants to reduce SQL queries, because `.browse()` automatically sets a new prefetch object if it doesn't get another one explicitly.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (yajo) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py]: TIL. It's cool!

```py
>>> d=collections.defaultdict(lambda: 0)
>>> d[1]+=5
>>> d[1]
5
```

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (pedrobaeza) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:133]: This is a special case that has been discussed a lot of times. This represents something similar as a class in Odoo context, and it's used by Odoo itself this way.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (pedrobaeza) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:164]: I'm not sure that's a good example, as dependent values are going to be recomputed each time you assign the variable. Another option can be to enclose everything on a `with self.env.norecompute():` clause, but current option is not bad at all.

**Comment by pedrobaeza** [comment]
Suggestions made or answered.

**Review by MiquelRForgeFlow — APPROVED** [review]
If this doesn't break anything, then it's fine for me.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:133]: If you like a new special case you should added to our guidelines.
For now, our guidelines of the OCA (not odoo) is use CamelCase for class name and snake_case for variables, uppercase for global variable.
If odoo uses other standard that is the odoo standard.
If you want join the odoo and oca standard and avoid discuss too many times you need added to the guidelines document.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:164]: if you uses “move.purchase_line_id[“qty_returned”] += qty” this is not re-comouted
It is just updating the cache value and you avoid an extra loop-for unneeded.

It just requires at start a self.update(qty_returned: 0) for the case where one record is not found.

**Comment by pedrobaeza** [comment]
> If this doesn't break anything, then it's fine for me.

Tests are there for checking this :wink:

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (pedrobaeza) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:133]: It's not a special case, as this is not the variable mentioned in the guidelines, and the linter is not forcing us to do the way you want. Open if you want the discussion for forbidding it and then I won't use it if approved.

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (pedrobaeza) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:164]: I'm not sure about that and it's difficult for me to check right now, but hey, that's not so important. That loop is very cheap in performance terms, so I prefer to keep it. Remember the sentence about performance:
```
Programmers waste enormous amounts of time thinking about, or worrying about, the 
speed of noncritical parts of their programs, and these attempts at efficiency
actually have a strong negative impact when debugging and maintenance are considered.

We should forget about small efficiencies, say about 97% of the time: premature optimization
is the root of all evil. Yet we should not pass up our opportunities in that critical 3%.
```

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:141]: I got it.
He is using a new browse for each iteration of the loop and he needs to re-use the same prefetch with all ids.

Thanks @Yajo to clarify.

**Review by moylop260 — APPROVED** [review]
Just small comments but not blocking.

**Comment by pedrobaeza** [comment]
@moylop260 although we don't agree on that small things that can be discussed at other level, they are not blocking for the merging, so I proceed for continuing my flow with the migration to v11 (you can review it as well on #396 :wink: ).

Thanks for your review (and the others' ones as well).

File: purchase_stock_picking_return_invoicing/models/purchase_order.py
Comment (moylop260) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:133]: Your first comment:
> This is a special case...

Your second comment:
> It's not a special case...

I think that you are contradicting the guidelines and your self comments.

But I don't have problem if you want skip our guideline [variable-names](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#variable-names) for now. If in the future you decide added it to avoid discussed it again and again it is good for me too.

I just approved this one.
Thanks for the great work here.

Comment (pedrobaeza) [reply purchase_stock_picking_return_invoicing/models/purchase_order.py:133]: Well, I haven't used the proper words, but they were following your comments. Any way, what I meant to say basically is: "this one is out of the guidelines". Call is "special case" or not. Remember that I was answering your sentence:

> If you like a new special case you should added to our guidelines.

So the proper answer would be "this is not the kind of special case inside the rule, as this can't be considered a normal variable (and thus is a special variable case)". Better?

I don't want to skip things, simply that they are not covered, but you know me, we can discuss what you want for generalizing this rule, but I don't want to get blocked this PR for that reason, and this want blocks me a huge migration to the the performance question.

**Pull request merged**
